code-review-chat/CodeReviewChat.ts

```diff
@@ -0,0 +1,123 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Octokit } from '@octokit/rest';
+import { WebClient } from '@slack/web-api';
+import { GitHubIssue } from '../api/api';
+// import { BlobServiceClient } from '@azure/storage-blob';
+import { safeLog } from '../common/utils';
+
+interface PR {
+	number: number;
+	body: string;
+	additions: number;
+	deletions: number;
+	changed_files: number;
+	url: string;
+	owner: string;
+	draft: boolean;
+}
+
+export interface Options {
+	slackToken: string;
+	storageConnectionString: string;
+	codereviewChannel: string;
+	payload: {
+		owner: string;
+		repo: string;
+		pr: PR;
+	};
+}
+
+export class BuildChat {
+	private pr: PR;
+	constructor(private octokit: Octokit, private issue: GitHubIssue, private options: Options) {
+		this.pr = options.payload.pr;
+	}
+
+	private async postMessage(message: string) {
+		const web = new WebClient(this.options.slackToken);
+		const memberships = await listAllMemberships(web);
+
+		const codereviewChannel =
+			this.options.codereviewChannel &&
+			memberships.find((m) => m.name === this.options.codereviewChannel);
+
+		if (!codereviewChannel) {
+			throw Error(`Slack channel not found: ${this.options.codereviewChannel}`);
+		}
+
+		await web.chat.postMessage({
+			text: message,
+			link_names: true,
+			channel: codereviewChannel.id,
+		});
+	}
+
+	async run() {
+		if (this.pr.draft) {
+			safeLog('PR is draft, ignoring');
+			return;
+		}
+
+		const data = await this.issue.getIssue();
+		const author = data.author;
+		if (!(await this.issue.hasWriteAccess(author))) {
+			safeLog('Issue author not team member, ignoring');
+			return;
+		}
+		if (!data.assignee) {
+			await this.issue.addAssignee(author.name);
+		}
+		const currentMilestone = await this.issue.getCurrentRepoMilestone();
+		if (!data.milestone && currentMilestone) {
+			await this.issue.setMilestone(currentMilestone);
+		}
+
+		const existing = await this.octokit.pulls.listReviewRequests({
+			owner: this.options.payload.owner,
+			repo: this.options.payload.repo,
+			pull_number: this.options.payload.pr.number,
+		});
+		const hasRequests = existing?.data?.users?.length;
+		if (hasRequests) {
+			safeLog('had existing review requests, exiting');
+			return;
+		}
+		const changedFilesMessage = `${this.pr.changed_files} file` + (this.pr.changed_files > 1 ? 's' : '');
+		const message = `${this.pr.owner}
```

Comment:
Any interest in using block kit for fancier messages? https://api.slack.com/block-kit

Reply:
At some point sure, just trying to get an initial POC working.